cirq-core/cirq/json_resolver_cache.py

```diff
@@ -54,6 +54,16 @@ def two_qubit_matrix_gate(matrix):
     def _parallel_gate_op(gate, qubits):
         return cirq.parallel_gate_op(gate, *qubits)
 
+    def _datetime(timestamp: float) -> datetime.datetime:
+        # As part of our serialization logic, we make sure we only serialize "aware"
+        # datetimes with the UTC timezone, so we implicitly add back in the UTC timezone here.
+        #
+        # Please note: even if the assumption is somehow violated, the fact that we use
+        # unix timestamps should mean that the deserialized datetime should refer to the
+        # same point in time but may not satisfy o = read_json(to_json(o)) because the actual
+        # timezones, and hour fields will not be identical.
+        return datetime.datetime.fromtimestamp(timestamp, tz=datetime.timezone.utc)
```

Comment:
yes, the docs for `utcnow` has a red box telling you not to use it.

> naive datetimes is simpler, as long as you use timestamp() for any kind of comparison

I figured it might be worth forcing folks to use the explicit thing so they can't accidentally not use `timestamp()`. I mean: another alternative would be to use timestamp floats instead of `datetime.datetime` everywhere.

the other problem is that it's really hard to use `datetime.__eq__` in a meaningful way with naive datetimes. I guess that falls under "comparison".

Reply:
Naive datetimes compare fine to other naive datetimes, and I think I'd lean toward using them when deserializing for simplicity. But this is definitely an area where footguns abound. We have switched over to using a `Timestamp` internally that is just a wrapper around unix epoch timestamps to avoid all these issues with `datetime`.

Reply:
> More explicit than timestamp floats would be timestamp protos, maybe made easier by `proto-plus` utilities? I personally prefer that, but my hunch is that it's better for the library to support `datetime` natively.
> 
> For usability of `datetime`, +1 to forcing users to use `.timestamp()` by setting the explicit tz.

You can call `.timestamp()` on any datetime instance, whether it has a tz or not, so I don't think that setting it forces users to do the right thing.

I think timestamp protos are going to feel out of place in most python code and we don't currently pull in protobuf as a dependency of `cirq-core`, so probably shouldn't rely on them unless we want to support time serialization only in `cirq-google`.

Reply:
on the basis that:

 1) the proto fields use an analog of utc "aware" datetimes
 2) things seem generally less broken if you live entirely in "aware" datetime land

I'd like to go ahead with forcing 'aware' during serialization and always deserializing as "aware" utc.

Reply:
Proto uses a timestamp (which can be gotten from naive or aware datetime), so I don't think that carries much weight here. But I'm fine with deserializing as aware and and accepting only aware datetimes (could decide to accept naive datetimes in the future if we want). Thanks for working through all the comments here!